runtime-sdk/modules/evm/src/precompile/subcall.rs

```diff
@@ -0,0 +1,521 @@
+use ethabi::{ParamType, Token};
+use evm::{
+    executor::stack::{PrecompileFailure, PrecompileHandle, PrecompileOutput},
+    ExitError, ExitSucceed,
+};
+
+use crate::backend::EVMBackendExt;
+use oasis_runtime_sdk::{
+    module::CallResult, modules::core::Error, subcall, types::transaction::CallerAddress,
+};
+
+use super::{record_linear_cost, PrecompileResult};
+
+/// A subcall validator which prevents any subcalls from re-entering the EVM module.
+struct ForbidReentrancy;
+
+impl subcall::Validator for ForbidReentrancy {
+    fn validate(&self, info: &subcall::SubcallInfo) -> Result<(), Error> {
+        if info.method.starts_with("evm.") {
+            return Err(Error::Forbidden);
+        }
+        Ok(())
+    }
+}
+
+const SUBCALL_BASE_COST: u64 = 10;
+const SUBCALL_WORD_COST: u64 = 1;
+
+pub(super) fn call_subcall<B: EVMBackendExt>(
+    handle: &mut impl PrecompileHandle,
+    backend: &B,
+) -> PrecompileResult {
+    record_linear_cost(
+        handle,
+        handle.input().len() as u64,
+        SUBCALL_BASE_COST,
+        SUBCALL_WORD_COST,
+    )?;
+
+    // Ensure that the precompile is called using a regular call (and not a delegatecall) so the
```

Comment:
Yes, but the same check handles that as well, it is just not explicitly mentioned.

Reply:
I thought that `callcode` doesn't use the address of the OAM, just `delegatecall`.

Reply:
You can check [here](https://github.com/rust-blockchain/evm/blob/b7b82c7e1fc57b7449d6dfa6826600de37cc1e65/runtime/src/eval/system.rs#L374-L390), so only a regular `call`/`staticcall` would set the `address` in EVM context to the address of the called contract (precompile). So the check will only pass for this case and not any of the others.

Obviously we could support other invocations as well by doing proper remapping (e.g. looking at `address` instead of `caller` in that case), but not sure if this is something that makes sense and it just results in more complexity.